File: go/types/token_store.go
package types

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

type storedToken struct {
	Token     string `json:"token"`
	ExpiresAt string `json:"expires_at"`
	IV        string `json:"iv"`
	Tag       string `json:"tag"`
}

type EncryptedTokenStore struct {
	filePath string
	key      []byte
}

func NewEncryptedTokenStore(filePath, encryptionKey string) *EncryptedTokenStore {
	key := []byte(encryptionKey)
	if len(key) < 32 {
		padded := make([]byte, 32)
		copy(padded, key)
		for i := len(key); i < 32; i++ {
			padded[i] = 'x'
		}
		key = padded
	}
	return &EncryptedTokenStore{filePath: filePath, key: key[:32]}
}

func (s *EncryptedTokenStore) Save(token string, expiresAt time.Time) error {
	iv := make([]byte, 12)
	if _, err := rand.Read(iv); err != nil {
		return err
	}

	block, err := aes.NewCipher(s.key)
	if err != nil {
		return err
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return err
	}

	encrypted := aesgcm.Seal(nil, iv, []byte(token), nil)

	dir := filepath.Dir(s.filePath)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return err
	}

	data := storedToken{
		Token:     string(encrypted),
		ExpiresAt: expiresAt.UTC().Format(time.RFC3339),
		IV:        string(iv),
		Tag:       "",
	}

	raw, err := json.Marshal(data)
	if err != nil {
		return err
	}

	return os.WriteFile(s.filePath, raw, 0600)
}

func (s *EncryptedTokenStore) Load() (string, time.Time, error) {
	raw, err := os.ReadFile(s.filePath)
	if err != nil {
		return "", time.Time{}, err
	}

	var data storedToken
	if err := json.Unmarshal(raw, &data); err != nil {
		return "", time.Time{}, err
	}

	block, err := aes.NewCipher(s.key)
	if err != nil {
		return "", time.Time{}, err
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", time.Time{}, err
	}

	iv := []byte(data.IV)
	decrypted, err := aesgcm.Open(nil, iv, []byte(data.Token), nil)
	if err != nil {
		return "", time.Time{}, err
	}

	expiresAt, err := time.Parse(time.RFC3339, data.ExpiresAt)
	if err != nil {
		return "", time.Time{}, err
	}

	return string(decrypted), expiresAt, nil
}

func (s *EncryptedTokenStore) Clear() error {
	return os.Remove(s.filePath)
}

File: typescript/src/utils/token-store.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";
import { readFileSync, writeFileSync, existsSync, mkdirSync } from "node:fs";

const ALGORITHM = "aes-256-gcm";

export interface TokenStoreConfig {
  filePath: string;
  encryptionKey: string;
}

interface StoredToken {
  token: string;
  expiresAt: string;
  iv: string;
  tag: string;
}

export class EncryptedTokenStore {
  private readonly filePath: string;
  private readonly key: Buffer;

  constructor(config: TokenStoreConfig) {
    this.filePath = config.filePath;
    this.key = Buffer.from(config.encryptionKey.padEnd(32, "x").slice(0, 32));
  }

  save(token: string, expiresAt: Date): void {
    const iv = randomBytes(16);
    const cipher = createCipheriv(ALGORITHM, this.key, iv);
    let encrypted = cipher.update(token, "utf8", "hex");
    encrypted += cipher.final("hex");
    const tag = cipher.getAuthTag().toString("hex");

    const dir = this.filePath.substring(0, this.filePath.lastIndexOf("/"));
    if (!existsSync(dir)) {
      mkdirSync(dir, { recursive: true });
    }

    const stored: StoredToken = {
      token: encrypted,
      expiresAt: expiresAt.toISOString(),
      iv: iv.toString("hex"),
      tag,
    };

    writeFileSync(this.filePath, JSON.stringify(stored), { mode: 0o600 });
  }

  load(): { token: string; expiresAt: Date } | null {
    if (!existsSync(this.filePath)) return null;

    try {
      const raw = readFileSync(this.filePath, "utf8");
      const stored: StoredToken = JSON.parse(raw);

      const decipher = createDecipheriv(
        ALGORITHM,
        this.key,
        Buffer.from(stored.iv, "hex"),
      );
      decipher.setAuthTag(Buffer.from(stored.tag, "hex"));
      let decrypted = decipher.update(stored.token, "hex", "utf8");
      decrypted += decipher.final("utf8");

      return {
        token: decrypted,
        expiresAt: new Date(stored.expiresAt),
      };
    } catch {
      return null;
    }
  }

  clear(): void {
    try {
      if (existsSync(this.filePath)) {
        writeFileSync(this.filePath, "");
      }
    } catch {
      // ignore
    }
  }
}
